pkg/txhandler-factory/simple/simple_transaction_handler.go

```diff
@@ -0,0 +1,457 @@
+	tx, err := sth.toolkit.Persistence.GetTransactionByID(ctx, txID)
+	if err != nil {
+		return nil, err
+	}
+	if tx == nil {
+		return nil, i18n.NewError(ctx, tmmsgs.MsgTransactionNotFound, txID)
+	}
+	return tx, nil
+}
+func (sth *simpleTransactionHandler) GetTransactions(ctx context.Context, afterStr, signer string, pending bool, limit int, direction toolkit.SortDirection) (transactions []*apitypes.ManagedTX, err error) {
+	var afterTx *apitypes.ManagedTX
+	if afterStr != "" {
+		// Get the transaction, as we need this to exist to pick the right field depending on the index that's been chosen
+		afterTx, err = sth.toolkit.Persistence.GetTransactionByID(ctx, afterStr)
+		if err != nil {
+			return nil, err
+		}
+		if afterTx == nil {
+			return nil, i18n.NewError(ctx, tmmsgs.MsgPaginationErrTxNotFound, afterStr)
+		}
+	}
+	switch {
+	case signer != "" && pending:
+		return nil, i18n.NewError(ctx, tmmsgs.MsgTXConflictSignerPending)
+	case signer != "":
+		var afterNonce *fftypes.FFBigInt
+		if afterTx != nil {
+			afterNonce = afterTx.Nonce
+		}
+		return sth.toolkit.Persistence.ListTransactionsByNonce(ctx, signer, afterNonce, limit, direction)
+	case pending:
+		var afterSequence *fftypes.UUID
+		if afterTx != nil {
+			afterSequence = afterTx.SequenceID
+		}
+		return sth.toolkit.Persistence.ListTransactionsPending(ctx, afterSequence, limit, direction)
+	default:
+		return sth.toolkit.Persistence.ListTransactionsByCreateTime(ctx, afterTx, limit, direction)
+	}
+}
+
+func (sth *simpleTransactionHandler) createManagedTx(ctx context.Context, txID string, txHeaders *ffcapi.TransactionHeaders, gas *fftypes.FFBigInt, transactionData string) (*apitypes.ManagedTX, error) {
+
+	// The request ID is the primary ID, and should be supplied by the user for idempotence
+	if txID == "" {
+		txID = fftypes.NewUUID().String()
+	}
+
+	// First job is to assign the next nonce to this request.
+	// We block any further sends on this nonce until we've got this one successfully into the node, or
+	// fail deterministically in a way that allows us to return it.
+	lockedNonce, err := sth.assignAndLockNonce(ctx, txID, txHeaders.From)
+	if err != nil {
+		return nil, err
+	}
+	// We will call markSpent() once we reach the point the nonce has been used
+	defer lockedNonce.complete(ctx)
+
+	// Sequencing ID is always generated by us - so we have a deterministic order of transactions
+	// Note: We must allocate this within the nonce lock, to ensure that the nonce sequence and the
+	//       global transaction sequence line up.
+	seqID := apitypes.NewULID()
+
+	// Next we update FireFly core with the pre-submitted record pending record, with the allocated nonce.
+	// From this point on, we will guide this transaction through to submission.
+	// We return an "ack" at this point, and dispatch the work of getting the transaction submitted
+	// to the background worker.
+	now := fftypes.Now()
+	mtx := &apitypes.ManagedTX{
```

Comment:
Ref earlier comments regarding the contract between the fields a TXHandler is responsible for and those FFTM is responsible for, I think this code really shows it.

We're passing to the TXHandler full responsibility to construct and persist the object, including fields that code completely outside of the TXHandler interface is responsible for.

I wonder if an `interface` with setters for the fields a TXHandler is allowed to change makes sense, and that's what the `toolkit.TXHistory` and `toolkit.Persistence` functions take in.

Then the instance is already pre-constructed as the input to `RegisterNewTransaction`, with a clearly documented description of what can be expected to be filled in already for you.

Reply:
I've removed the `previousTransactionHash` field from the object to avoid polluting it with data that shouldn't be part of the persisted object.

In terms of those history fields on the object, we can discuss and split them in a separate PR.

Reply:
Capturing some discussions around this based on my understanding. `ManagedTx` serves many different purposes today, here is a list I collected through conversations. Listing them out to figure out whether there are some understanding gaps and to figure out the key motivation for splitting the ManagedTx object.

Shared by different levels:
1. Defines the REST API object schema for transactions resource.
2. Used by transaction handler to record state information for asynchronous processing (enabling the processing logic to be stateless).
3. Used by Persistence service as the schema of the data entries.



Shared by different components:
1. For storing immutable transaction data to represent transaction request record
2. For storing mutable transactions data that can change based on the handling logic of transaction preparation and submission
3. For storing blockchain receipts and confirmations for transaction tracking to completion
4. Used by History service to store histories of transaction processing.